feat: Implement "Async" RPC

L3tum opened the pull request:
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌ 
| New feature?  | ✔️ TODO please update "Other Features" section in CHANGELOG.md file
| Issues        | https://github.com/orgs/roadrunner-server/discussions/1850 
| Docs PR       | spiral/docs#... <!-- prefix each issue number with "spiral/docs#", required only for new features --> TODO?

Implements an "Async" RPC Interface and Implementation using multiple relays to effectively offer non-blocking IO in regards to the Roadrunner communication.

As discussed in the linked discussion, there's a (small but signifcant) delay in the RPC communication to Roadrunner that *can* have an impact on a service. In my case the service had a nominal response time of ~2ms, went up to 3ms with metric collection and also had a 2ms `kernel.terminate` listener to collect further metrics, so all in all was busy for ~5ms with only ~2ms spent on the actual request->response cycle.

As a fix I propose an "async" implementation of the RPC interface. I don't really like calling it async, because it doesn't offer the same versatility as a proper async implementation, but it changes the control flow and thus qualifies for that name, I guess.
The interface has 3 significant new methods as well as a few supporting changes in other code:
- callIgnoreResponse => Calls a Roadrunner method but doesn't wait for the response and ignores it if it's received. This is intended for "best-effort" areas such as Metrics collection, where it is generally accepted if some Metrics get lost on the way
- callAsync => Calls a Roadrunner method and tracks the response rather than wait for it. The response can then be received with...
- getResponse => Gets and if needed waits for a response from a previous call to callAsync. The request/response is tracked with the already used $seq. The implementation keeps the response in the socket buffer as long as possible until it's needed to reduce unnecessary delays
- The supporting methods `hasResponse` and `hasAnyResponse` just check if a response has been received

**Results**
I've measured the impact in a testcase using the same repro used in the discussion. The results are the following:
- Sync Metrics Call: 1.228ms
- AlwaysIgnoreResponseRPC (from the repro) Call: 0.1366ms
- MultiRPC Call (this change): 0.1399ms
As you can see the "good" implementation (this change) is within the normal deviation of the always-ignore implementation and offers an improvement ~10x performance. I've been using a (slightly modified) implementation of this in the aforementioned production service and could cut down the time spent collecting metrics from a total of 3ms to just 0,2ms, with a min/max of respectively 0.1ms and 0.3ms.

I've also added a change (PR following) in roadrunner-kv that uses the `callAsync` method to implement a `setAsync` function. The results so far are promising. With a follow-up check on whether the `kv.Set` call was successful, the async implementation is significantly faster:

- "SyncKVImpactCacheStartup": "56.2324ms"
- "SyncKVImpactNewItem": "0.5328ms"
- "SyncKVImpactExistingItemSameValue: "0.2113ms"
- "SyncKVImpactExistingItemNewValue": "0.2006ms"
- "AsyncKVImpactCacheStartup": "21.7821ms"
- "AsyncKVImpactNewItem": "0.1146ms"

Full results for the Metrics calls:
````json
{
    "Base": {
        "Max": "0.2063ms",
        "Min": "0.0954ms",
        "Avg": "0.1228ms",
        "Total": "1.228ms"
    },
    "AlwaysIgnoreResponseRPC": {
        "Max": "0.0537ms",
        "Min": "0.0069ms",
        "Avg": "0.01366ms",
        "Total": "0.1366ms"
    },
    "MultiRPC": {
        "Max": "0.0714ms",
        "Min": "0.0061ms",
        "Avg": "0.01399ms",
        "Total": "0.1399ms"
    }
}
````


**Considerations**
- The whole implementation is "opt-in", which can also be seen in my two follow-up PRs ([#1](https://github.com/roadrunner-php/metrics/pull/13) and [#2](https://github.com/roadrunner-php/kv/pull/36)). I didn't want to force it in case someone depends on the existing implementation. It makes some of the code look a bit messy due to having to support both code paths but I think it's the better solution. I'm open to change it though if you disagree.
- I had to "open" two previously private fields in the Relay implementations to make more effective use of `socket_select` and `stream_select`. I've commented it as well as marked them with `@internal` but it's definitely not ideal.
- ~The overflow response buffer currently has no upper limit and thus it's somewhat easy to get a memory leak here. I could add a check in `callAsync` to see if the response buffer is above a certain limit (1000?) and if so flush it. I'll probably do so when adding the tests.~ Implemented a flush at 1000 entries (realistically the most a normal application should ever see is maybe ~50) -- I've upped this to 10_000 entries because in testing in particular Doctrine caches get hammered extremely hard and 1_000 is not enough.

**TODO**
- ~I have glanced at the tests in this repo and will try to implement a few for this change if you've given me the go-ahead with the current implementation.~ Added a number of tests
- Add a Changelog.md entry
- Add documentation for this (?)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
    - Introduced multi-relay functionality for enhanced message and relay management.
    - Added asynchronous RPC communication capabilities over multiple relays.
    - Improved RPC functionality with new methods and class structures for more efficient communication and coding practices.
- **Bug Fixes**
    - Corrected a typo in RPCInterface method description for clarity.
- **Refactor**
    - Updated the RPC class to extend from `AbstractRPC` with various method and property adjustments for streamlined operation.
    - Modified visibility of certain properties in `SocketRelay` and `StreamRelay` to public, with added documentation for these changes.
- **Tests**
    - Added a comprehensive test suite for multiple RPC interactions, including exception handling and asynchronous calls.
    - Introduced test cases specifically for MsgPack and TCP communication in the MultiRPC context.
- **Documentation**
    - Enhanced internal documentation to explain visibility changes in relay properties.
- **Chores**
    - Implemented minor server-side enhancements, including a new `SleepEcho` function for testing purposes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

L3tum committed:
feat: Implement "Async" RPC

L3tum committed:
fix: Issues uncovered by Psalm

L3tum committed:
feat: Add getResponses as well as fix a number of logic or Psalm errors

L3tum committed:
feat: Add an unholy amount of tests

L3tum committed:
feat: Optimize getResponse to remove array_search and fix potential opportunities to lose a relay

L3tum committed:
chore: Add test for response buffer handling

L3tum committed:
fix: Accidentally saved too many responses

L3tum committed:
fix: Reorder methods to make sure we do not lose a relay

L3tum committed:
fix: Wrong order for $seq

L3tum committed:
fix: Up response buffer maximum

L3tum committed:
fix: Add Error Handling to MultiRPC::getResponses()

L3tum committed:
feat: Simplify MultiRPC and MultiRelayHelper, fixes some issues resulting in lost relays (actually improves performance as well)

L3tum committed:
fix: Actually call tests

L3tum committed:
fix: Model array_key_last output as docblock

L3tum committed:
fix: Issues introduced by simplification of relay handling

L3tum commented:
Hey @msmakouz , I saw @rustatian 's comment, thank you for the review :)

I've just pushed some last changes that were needed after I tried using this implementation in my prod service. It didn't behave very well and once I started down that rabbit hole I realized I had made this more complicated than it really needed to. So I done away with the distinction between occupied relays with/out responses and also removed the limitation from the MultiRelayHelper of not supporting a map (aka non-int indexed array). I'm sorry if you had already started looking through the code, but from what I can tell after having spent a significant amount of time debugging the tests the whole thing should be pretty final now.

L3tum committed:
fix: Gracefully handle socket disconnect without blocking (too much) on it

L3tum committed:
fix: Handle cloning of MultiRPC

L3tum commented:
Hey @msmakouz , last (functional) change from my side (hopefully). I fell into the trap of cloning an array of objects, not realizing they get shared between the instances.
I experimented with making the arrays static instead, but it felt like a bit of a hack, especially in regards to fetching a response from the buffer of another RPC (thus running into codec issues, for example).
Overall I'm a bit disappointed Psalm didn't catch this

L3tum committed:
fix: Handle cloning of MultiRPC with SocketRelay

L3tum commented:
Hey @msmakouz / @rustatian , one thing I noticed now is that the number of Goroutines apparently explodes with this change. I'm not sure where or why because I'm honestly not knowledgable in Go at all. In my case I've noticed a growth from ~500 to ~2000. That's with 100 sockets per Worker with 8 Workers opened. So I guess there's a number of goroutines running per socket or something like that?

The issue is mostly during high load scenarios, cause it spawned up to 15k Goroutines until RR crashed completely.
They do get cleaned up afterwards (without a crash, I mean) so I don't think it'ss a leak.

If I understand Go correctly a socket there spawns two Goroutines (Read & Write), so at 100 sockets would be 200 Goroutines per Worker, with 8 Workers at 1600 Goroutines, so that makes sense. What I don't get is why there's 15k Goroutines when any type of load is applied

msmakouz commented:
@L3tum Hi, thanks for the work done! Do you plan to make any more changes to the code, or can we look at your proposed implementation?
At first glance, I see that there is insufficient unit test coverage, but I will first look at the implementation itself, if it is already ready.

L3tum commented:
Hey @msmakouz , so far I think it's about done. I'm not sure why it thinks it doesn't have any unit test coverage, as far as I can see there should be almost everything covered via all the tests in `tests/Goridge/MultiRPC`

L3tum commented:
Okay, @msmakouz undrafting the PR at least allowed the AI to go over it. Pretty good comments from it actually, didn't expect that. Last time I used something like that it only found the most obvious^^

I'll address most of its issues tonight or tomorrow. One thing, there's a few marks I'd like to tell it to "shut up", is there a way for that? For example complaining about missing error handling in the caller, when the caller declares the errors as `@throws` so doesn't need (or should) explicit handling. I'll check if I missed a throw somewhere if that's what the mark is about.

L3tum committed:
fix: Typo in Testclass Name

L3tum committed:
fix: Add comments documenting ensureFreeRelayAvailable and getResponseFromRelay

L3tum committed:
fix: Simplify getResponses() a bit

msmakouz commented:
> One thing, there's a few marks I'd like to tell it to "shut up", is there a way for that?

You can leave a comment and click the "Resolve conversation" button if you think the AI is wrong. If you don't have this button, then don't pay attention to this comment.

<img width="830" alt="33" src="https://github.com/roadrunner-php/goridge/assets/67324318/fc5e308a-571a-4e6c-b501-0a9be4baea75">

L3tum committed:
fix: Add configurable buffer threshold and change exception message

L3tum commented on tests/Goridge/MultiRPC.php:
```diff
@@ -0,0 +1,752 @@
+        $this->assertFreeRelaysCorrectNumber($this->rpc);
+    }
+
+    protected function assertFreeRelaysCorrectNumber(GoridgeMultiRPC $rpc): void
+    {
+        $property = new ReflectionProperty(GoridgeMultiRPC::class, 'freeRelays');
+        $numberOfFreeRelays = count($property->getValue($rpc));
+        $property = new ReflectionProperty(GoridgeMultiRPC::class, 'occupiedRelays');
+        $numberOfOccupiedRelays = count($property->getValue($rpc));
+        $property = new ReflectionProperty(GoridgeMultiRPC::class, 'seqToRelayMap');
+        $numberOfWaitingResponses = count($property->getValue($rpc));
+
+        $this->assertSame(
+            10,
+            $numberOfFreeRelays + $numberOfOccupiedRelays,
+            "RPC has lost at least one relay! Waiting Responses: $numberOfWaitingResponses, Free Relays: $numberOfFreeRelays, Occupied Relays: $numberOfOccupiedRelays"
+        );
+    }
```
This does not make sense. A sensible test would only be to check for `$numberOfWaitingResponses == 0`, but we may not have actually read through all expected responses at this point due to the nature of some of the tests. Thus we leave it to PHP to clean it up and don't assert any particular state except that all relays are accounted for.

L3tum commented on tests/Goridge/TPCMultiRPCTest.php:
```diff
@@ -0,0 +1,15 @@
+use Spiral\Goridge\SocketType;
+use Spiral\Goridge\Tests\MultiRPC;
+
+class TPCMultiRPCTest extends MultiRPC
```
Fixed in 45ef16a5

L3tum commented on tests/Goridge/MultiRPC.php:
```diff
@@ -0,0 +1,752 @@
+    public function testAllowsOnlySockets(): void{
+        $this->expectException(RPCException::class);
+        $this->expectExceptionMessage("MultiRPC can only be used with sockets, no pipes allowed");
+        $this->rpc = new GoridgeMultiRPC([new StreamRelay(STDIN, STDOUT)]);
```
I pushed a fix in commit 825953a8

L3tum replied:
Sorry, wrong branch, I pushed a fix in commit 825953a80d3e299fe688e4a25286a09d89c37f7c

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+        foreach ($this->occupiedRelays as $key => $relay) {
+            $this->occupiedRelays[$key] = clone $relay;
+        }
+        $this->seqToRelayMap = $this->asyncResponseBuffer = [];
```
I pushed a fix in commit 825953a8

L3tum replied:
Please review this comment again

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+            $this->asyncResponseBuffer[$expectedSeq] = $frame;
+        }
+
+        return $frame;
```
I pushed a fix in commit 4a46e785

L3tum replied:
Please review this comment again

L3tum replied:
I pushed a fix in commit 4a46e785ed80c8419697f8172c8049c5b9692c7c

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+        // Sometimes check if all occupied relays are even still connected
+        $this->checkAllOccupiedRelaysStillConnected();
+
+        return array_key_last($this->freeRelays);
```
I pushed a fix in commit 4a46e785

L3tum replied:
Please review this comment again

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+
+                unset($seqsToRelays[$seq], $this->seqToRelayMap[$seq]);
+            }
+        }
```
I pushed a fix in 2cb78fb2

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+            $frame = $this->getResponseFromRelay($relay, $seq, true);
+        }
+
+        return $this->decodeResponse($frame, $relay, $options);
```
I'm open for suggestions (@msmakouz )

L3tum replied:
I've pushed a fix in commit aadfe5b2

L3tum replied:
I've pushed a fix in commit 33b234a2

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+            return true;
+        }
+
+        return false;
```
I've pushed a fix in commit 825953a8

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+        // Last index so no need for array_pop or stuff
+        unset($this->freeRelays[$relayIndex]);
+
+        return $seq;
```
I've pushed a fix in commit 825953a8

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+        return $this->decodeResponse($frame, $relay, $options);
+    }
+
+    public function callIgnoreResponse(string $method, mixed $payload): void
+    {
+        $relayIndex = $this->ensureFreeRelayAvailable();
+        $relay = $this->freeRelays[$relayIndex];
+
+        $relay->send($this->packFrame($method, $payload));
+
+        $seq = self::$seq;
+        self::$seq++;
+        $this->occupiedRelays[$seq] = $relay;
+        // Last index so no need for array_pop or stuff
+        unset($this->freeRelays[$relayIndex]);
```
I think it complains about the declared `@throws \Error` in SocketRelay::connect().
Afaict that Error should never be thrown since `extension_loaded(sockets)` is checked in the constructor. Lmk if I can remove that

L3tum replied:
@coderabbitai Are these runtime errors not covered by `socket_last_error()`? The documentation for `socket_connect()` does not imply that it would throw an Error.

L3tum replied:
@msmakouz Following this exchange and reading the docs more I've removed the annotation in commit c4b41cff

L3tum committed:
feat: Refactor code around specialty handling of SocketRelay and add more test coverage

L3tum committed:
fix: Make exception more descriptive

L3tum committed:
fix: Missing extends statement in interface

L3tum committed:
fix: Enforce __clone impl and test streams with data on them

L3tum commented on src/ConnectedRelayInterface.php:
```diff
@@ -0,0 +1,30 @@
+use Spiral\Goridge\Exception\RelayException;
+
+/**
+ * This interface describes a Relay that explictily establishes a connection.
+ * That connection can also be re-established on the fly (in comparison to StreamRelay, which relies on the existence of the streams).
+ * The object is also clonable, i.e. supports cloning without data errors due to shared state.
```
I've pushed a fix in commit 0514503

L3tum commented on tests/Goridge/MultiRelayHelperTest.php:
```diff
@@ -0,0 +1,17 @@
+
+class MultiRelayHelperTest extends TestCase
+{
+    public function testSupportsStreamRelay(): void
+    {
+        $relays = [new StreamRelay(STDIN, STDOUT), new StreamRelay(STDIN, STDERR)];
+        // No message available on STDIN, aka a read would block, so this returns false
+        $this->assertFalse(MultiRelayHelper::findRelayWithMessage($relays));
+    }
```
Fixed in 0514503

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+    ) {
+        foreach ($relays as $relay) {
+            if (!($relay instanceof SocketRelay)) {
+                throw new RPCException("MultiRPC can only be used with sockets, no pipes allowed");
```
I've pushed a fix in commit ca811f76

L3tum replied:
I've pushed a fix in commit bb190ce8

L3tum committed:
fix: Remove @throws Error annotation

L3tum commented on src/RPC/MultiRPC.php:
```diff
@@ -0,0 +1,404 @@
+            // Force connect
+            $relay->connect();
+        }
+    }
```
IMHO if `connect()` encounters an issue the exception should bubble up ASAP cause otherwise the error would only surface in the first call to a `call*` method, while `preConnectRelays` is likely already called on instantiation.

L3tum commented:
Hey @msmakouz !

I think I implemented all suggestions by the AI and now I'm only massaging it to re-check, but I don't expect any further issues to pop up. In case you waited for this to review the changes yourself.

L3tum commented:
@coderabbitai review

L3tum commented on tests/Goridge/MultiRelayHelperTest.php:
```diff
@@ -0,0 +1,27 @@
+
+class MultiRelayHelperTest extends TestCase
+{
+    public function testSupportsStreamRelay(): void
+    {
+        $relays = [new StreamRelay(STDIN, STDOUT), new StreamRelay(STDIN, STDERR)];
+        // No message available on STDIN, aka a read would block, so this returns false
+        $this->assertFalse(MultiRelayHelper::findRelayWithMessage($relays));
+    }
```
Message availability is tested in the test below this one.

L3tum commented on tests/Goridge/MultiRelayHelperTest.php:
```diff
@@ -0,0 +1,27 @@
+        $this->assertFalse(MultiRelayHelper::findRelayWithMessage($relays));
+    }
+
+    public function testSupportsReadingFromStreamRelay(): void
+    {
+        $stream = fopen('php://temp', 'rw+');
+        fwrite($stream, 'Hello');
+        fseek($stream, 0);
+        $relays = [new StreamRelay($stream, STDOUT)];
+        $this->assertCount(1, MultiRelayHelper::findRelayWithMessage($relays));
+        fclose($stream);
+    }
```
Reading from a stream is covered by php-src.

roxblnfk committed:
Fix code style

roxblnfk commented:
Hello.

I think the API can be greatly simplified if we introduce a `DeferredResult` object. This object would be returned from the `callAsync()` method. If the user is interested in the result, they will use `DeferredResult`. If not, the object will be naturally deleted after the method is executed.

In addition, I would make all RPC asynchronous right away, then the call to the synchronous method `call()` would look like `callAsync()->await()`.

L3tum commented:
> Hello.
> 
> I think the API can be greatly simplified if we introduce a `DeferredResult` object. This object would be returned from the `callAsync()` method. If the user is interested in the result, they will use `DeferredResult`. If not, the object will be naturally deleted after the method is executed.
> 
> ```
> // an example
> interface DeferredResult
> {
>     /**
>      * Check whether the request is pending.
>      */
>     public function isPending(): bool;
> 
>     /**
>      * Await and return the result. Throws an exception if the call failed.s
>      *
>      * @throws GoridgeException
>      * @throws RPCException
>      * @throws ServiceException
>      */
>     public function getResult(): mixed;
> }
> ```
> 
> In addition, I would make all RPC asynchronous right away, then the call to the synchronous method `call()` would look like `callAsync()->await()`.

Hey @roxblnfk , sure that looks good. The question is whether we want to use something custom or use either a `Promise` library so it can be adapted into other frameworks, or even `Fiber` directly. We'd have to put some logic into its destructor though to manage the state in `MultiRPC`, since it would need to delete its entry in `seqToRelayMap` at least. It would stretch out some of the code and simplify other parts so overall it'd be a good change I think.

L3tum commented:
I've used Fibers for that [in this repo](https://github.com/L3tum/php-uring-server/blob/master/src/FiberServer.php#L88). A fiber is in the end just a promise so we can use it as a promise.
Regarding using a promise lib, the idea behind that is to enable interop with all the frameworks that roadrunner supports. We'd still write our own `Promise`, but the shared interface would enable it to work with others

L3tum commented:
In that case I would argue that an abstraction such as `DeferredResult` (or w/e) just adds overhead when it isn't needed. In my tests the repeated object instantiation alone can make up around 20% of performance regression. And adding an object pool just isn't really practical for that type of object in PHP, since we can't access the reference count.

Here's the patch for the change [1]. As you can see there really isn't that much actual change in there, and arguably no reduction of complexity. The biggest code change are the changes to `hasResponses` and `getResponses`, which would be kind of awkward with this API. The simplification of the three `call*` methods is nice, but there's a reason why they are separate, since unifying them also adds a few unnecessary array operations, which are pretty bad for performance.

[1]
````diff
Index: src/RPC/MultiRPC.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/RPC/MultiRPC.php b/src/RPC/MultiRPC.php
--- a/src/RPC/MultiRPC.php	(revision c4b41cff3d9269ae999aedb4e206c596a99690eb)
+++ b/src/RPC/MultiRPC.php	(date 1708624735013)
@@ -139,34 +139,15 @@
 
     public function call(string $method, mixed $payload, mixed $options = null): mixed
     {
-        $relayIndex = $this->ensureFreeRelayAvailable();
-        $relay = $this->freeRelays[$relayIndex];
-
-        $relay->send($this->packFrame($method, $payload));
-
-        // wait for the frame confirmation
-        $frame = $this->getResponseFromRelay($relay, self::$seq, true);
-
-        self::$seq++;
-
-        return $this->decodeResponse($frame, $relay, $options);
+        return $this->callAsync($method, $payload)->getResult($options);
     }
 
     public function callIgnoreResponse(string $method, mixed $payload): void
     {
-        $relayIndex = $this->ensureFreeRelayAvailable();
-        $relay = $this->freeRelays[$relayIndex];
-
-        $relay->send($this->packFrame($method, $payload));
-
-        $seq = self::$seq;
-        self::$seq++;
-        $this->occupiedRelays[$seq] = $relay;
-        // Last index so no need for array_pop or stuff
-        unset($this->freeRelays[$relayIndex]);
+        $this->callAsync($method, $payload)->ignore();
     }
 
-    public function callAsync(string $method, mixed $payload): int
+    public function callAsync(string $method, mixed $payload): DeferredResult
     {
         // Flush buffer if someone doesn't call getResponse
         if (count($this->asyncResponseBuffer) > $this->asyncBufferThreshold) {
@@ -188,7 +169,7 @@
         // Last index so no need for array_pop or stuff
         unset($this->freeRelays[$relayIndex]);
 
-        return $seq;
+        return new DeferredResult($this, $seq);
     }
 
     public function hasResponse(int $seq): bool
@@ -206,8 +187,20 @@
         return false;
     }
 
-    public function hasResponses(array $seqs): array
+    /**
+     * @param DeferredResult[] $results
+     * @return DeferredResult[]
+     */
+    public function hasResponses(array $results): array
     {
+        $seqMap = [];
+        $seqs = [];
+
+        foreach ($results as $result) {
+            $seqMap[$result->id] = $result;
+            $seqs[] = $result->id;
+        }
+
         $relays = [];
         /** @var array<int, positive-int> $relayIndexToSeq */
         $relayIndexToSeq = [];
@@ -233,7 +226,12 @@
             $seqsWithResponse[] = $relayIndexToSeq[$relayIndex];
         }
 
-        return $seqsWithResponse;
+        $resultsWithResponses = [];
+        foreach ($seqsWithResponse as $seq) {
+            $resultsWithResponses[] = $seqMap[$seq];
+        }
+
+        return $resultsWithResponses;
     }
 
     public function getResponse(int $seq, mixed $options = null): mixed
@@ -257,13 +255,15 @@
         return $this->decodeResponse($frame, $relay, $options);
     }
 
-    public function getResponses(array $seqs, mixed $options = null): iterable
+    public function getResponses(array $results, mixed $options = null): iterable
     {
         // Quick return
-        if (count($seqs) === 0) {
+        if (count($results) === 0) {
             return;
         }
 
+        $seqs = array_map(static fn(DeferredResult $result) => $result->id, $results);
+
         // Flip the array to use the $seqs for key indexing
         $seqsKeyed = [];
 

````

wolfy-j commented:
I would keep the original call method the same as before, to be honest. It's under the interface so we can always change to fiber friendly implementation later. 

For now, this PR looks good to me. Can we make sure that tests pass?

rustatian commented:
@L3tum Added you to the `contributors` team. You won't need to create a fork to push a PR to this repo. Please don't hesitate to ping me if you want to contribute more, so, I'll try you make your life with forks easier 😄

L3tum commented:
Thank you @rustatian !

I'll close this PR then and push a branch to open a new one. Would also remove the mess I made of the code review AI 😆 

@wolfy-j Sounds good to me. 

@msmakouz / @roxblnfk / @wolfy-j  I've fixed the tests and also fixed a bug related to `clone`ing the MultiRPC. In particular the issue stems from the usage of it in the `Cache` class in roadrunner-php/kv, but also others that (would) make use of it. While the cloning was handled alright, each clone would increase the number of relays created. In my prod service for example I've used 7 different cache services. The result was that each worker created 400 relays, at 12 workers per container so 4800 relays/sockets, with 2 goroutines per socket so 9600 goroutines. That exploded the memory usage to an insane ~1,5GB entirely in socket buffers.
Because of that I had to change a lot of tests, but the implementation itself just moved to `static` fields so I hope it doesn't add too much burden on you to review that. I'll tag you on the new PR once that is ready :)

L3tum commented:
I've moved the PR to https://github.com/roadrunner-php/goridge/pull/25 :)
